plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/KSerializerDescriptorResolver.kt

```diff
@@ -394,7 +401,7 @@ object KSerializerDescriptorResolver {
                 f.source)
         )
 
-        val kSerialClass = thisClass.getClassFromSerializationPackage("KSerializer").toSimpleType(false)
+        val kSerialClass = KotlinTypeFactory.simpleNotNullType(Annotations.EMPTY, thisClass.getClassFromSerializationPackage("KSerializer"), typeArguments)
```

Comment:
In general, this seems to be the right approach to fix issue, because `KSerializer` indeed has type arguments. However, you've created here one typearg `T`, and if your class will have more than one typearg, e.g. `Data<T0, T1>`, signature should look like `write$Self(KOutput, KSerialClassDesc, KSerializer<T0>, KSerializer<T1>`, not `...KSerializer<T>, KSerializer<T>)`. You can take a look at how `createSerializerGetterDescriptor` implemented, since it should has almost same signature.